pkg/kubelet/kubelet_pods.go

```diff
@@ -509,30 +509,35 @@ func (kl *Kubelet) getServiceEnvVarMap(ns string, enableServiceLinks bool) (map[
 		// Kubelets without masters (e.g. plain GCE ContainerVM) don't set env vars.
 		return m, nil
 	}
-	services, err := kl.serviceLister.List(labels.Everything())
-	if err != nil {
-		return m, fmt.Errorf("failed to list services when setting up env vars")
-	}
 
-	// project the services in namespace ns onto the master services
-	for i := range services {
-		service := services[i]
-		// ignore services where ClusterIP is "None" or empty
-		if !v1helper.IsServiceIPSet(service) {
-			continue
+	if enableServiceLinks {
+		services, err := kl.serviceLister.Services(ns).List(labels.Everything())
+		if err != nil {
+			return m, fmt.Errorf("failed to list services when setting up env vars")
 		}
-		serviceName := service.Name
 
-		// We always want to add environment variabled for master services
-		// from the master service namespace, even if enableServiceLinks is false.
-		// We also add environment variables for other services in the same
-		// namespace, if enableServiceLinks is true.
-		if service.Namespace == kl.masterServiceNamespace && masterServices.Has(serviceName) {
-			if _, exists := serviceMap[serviceName]; !exists {
-				serviceMap[serviceName] = service
+		// project the services in namespace ns onto the master services
+		for _, service := range services {
+			// ignore services where ClusterIP is "None" or empty
+			if !v1helper.IsServiceIPSet(service) {
+				continue
 			}
-		} else if service.Namespace == ns && enableServiceLinks {
-			serviceMap[serviceName] = service
+
+			serviceMap[service.Name] = service
+		}
+	}
+
+	// We always want to add environment variabled for master services
+	// from the master service namespace, even if enableServiceLinks is false.
+	// We also add environment variables for other services in the same
+	// namespace, if enableServiceLinks is true.
+	if _, hasMasterService := serviceMap[masterServiceName]; !hasMasterService {
+		masterService, err := kl.serviceLister.Services(kl.masterServiceNamespace).Get(masterServiceName)
+		if err != nil && !errors.IsNotFound(err) {
+			return m, fmt.Errorf("failed to get master service")
+		}
+		if err == nil {
```

Comment:
also need to replicate the `if !v1helper.IsServiceIPSet(service) {` skip here

Reply:
Thanks! @liggitt